Cross-compile RemoteExecutor

ViktorHofer opened the pull request:
Instead of using a single netstandard2.0 tfm and putting tfm specific stuff together (runtimeconfig.json which currently also gets deployed when referencing the package from netfx) cross-compiling both libraries. I still had to manually add a runtimeconfig.json as otherwise it wouldn't get deployed in the ProjectReference with PrivateAssets=All. Also I wasn't able to get the deps.json files also copied over.

@dasMulli any idea with PrivateAssets="All" on a ProjectReference doesn't include the $(TargetName).runtimeconfig.json and the $(TargetName).deps.json assets?

dasMulli commented:
The build and publish runtime configurations don't flow across P2Ps, they are part of the "PrepareForRun" step and note included in the targets that assemble the copy local items for P2Ps references.
When packing assemblies via P2Ps i previously resorted to packing publish outputs (see https://github.com/dotnet/sdk/issues/1846#issuecomment-356726385) or messing with project references so that they can forward their publish output.. see https://github.com/dasMulli/AssemblyInfoGenerationSdk/blob/master/src/DasMulli.AssemblyInfoGeneration.Sdk/sdk-layout.targets (the test project references the SDK project as P2P with a special property - which will do a publish and forward the publish result, with the added complexity of referencing the outer bulid of a multi-targeting project and forwarding a custom layout.. it is broken in a recent VS version but i think it only needs a minor change)

ViktorHofer commented:
@dasMulli thanks a lot for letting me know! Packaging the publish out is what we do in some other packages here in arcade. But we usually only do this for build tools that land in the tools directory. I think I will stick with the current solution.

@ericstj PTAL

ericstj commented on src/Microsoft.DotNet.RemoteExecutor/src/Microsoft.DotNet.RemoteExecutor/Microsoft.DotNet.RemoteExecutor.csproj:
```diff
@@ -1,7 +1,7 @@
 
   <PropertyGroup>
-    <TargetFramework>netstandard2.0</TargetFramework>
+    <TargetFrameworks>netcoreapp2.0;net472;netstandard2.0</TargetFrameworks>
```
Who uses the netstandard2.0 host?

ericstj commented on src/Microsoft.DotNet.RemoteExecutor/src/Microsoft.DotNet.RemoteExecutor/Microsoft.DotNet.RemoteExecutor.csproj:
```diff
@@ -23,8 +23,9 @@
     <ItemGroup>
-      <TfmSpecificPackageFile Include="$(OutputPath)Microsoft.DotNet.RemoteExecutorHost.dll" PackagePath="lib\$(TargetFramework)\" />
-      <TfmSpecificPackageFile Include="$(OutputPath)Microsoft.DotNet.RemoteExecutorHost.runtimeconfig.json" PackagePath="lib\$(TargetFramework)\" />
+      <TfmSpecificPackageFile Include="%(_ResolvedProjectReferencePaths.Identity)" PackagePath="lib\$(TargetFramework)\" />
```
It still feels unusual to me to put this under lib.  You actually don't want people to reference the host at all.

ViktorHofer replied:
@ericstj any idea how to solve this? otherwise I would be fine keeping the host in the lib folder next to the RemoteExecutor library.

ericstj replied:
A couple ways to solve this.  https://docs.microsoft.com/en-us/nuget/reference/nuspec#explicit-assembly-references, or treat it as content, or build folder + targets.

Barring all that, maybe just limit the public surface area and for anything that must be public, set EditorBrowsable.

ViktorHofer replied:
I misunderstood. I thought you don't like having the host library next to the lib library in the output directory. But you just don't like the way it is currently bundled in the nuget package. I will probably add a target with a CopyToOutputDirectory entry. Thanks!

ericstj replied:
> you just don't like the way it is currently bundled in the nuget package

Well, more specifically I don't like it being passed to the compiler.   I don't care what it looks like in the package.

ViktorHofer replied:
Updated the PR.

ViktorHofer commented:
@ericstj this is ready to go in. awaiting final review :)

ericstj commented on src/Microsoft.DotNet.RemoteExecutor/src/Microsoft.DotNet.RemoteExecutor/Microsoft.DotNet.RemoteExecutor.csproj:
```diff
@@ -1,30 +1,32 @@
     <ItemGroup>
-      <TfmSpecificPackageFile Include="$(OutputPath)Microsoft.DotNet.RemoteExecutorHost.dll" PackagePath="lib\$(TargetFramework)\" />
-      <TfmSpecificPackageFile Include="$(OutputPath)Microsoft.DotNet.RemoteExecutorHost.runtimeconfig.json" PackagePath="lib\$(TargetFramework)\" />
+      <TfmSpecificPackageFile Include="%(_ResolvedProjectReferencePaths.Identity)" PackagePath="tools\$(TargetFramework)\" />
```
Instead of using this private item, you could set OutputItemType on the projectreference.  https://blogs.msdn.microsoft.com/kirillosenkov/2015/04/04/how-to-have-a-project-reference-without-referencing-the-actual-binary/

ViktorHofer commented:
OK addressed your feedback. Package content:

![image](https://user-images.githubusercontent.com/7412651/56577892-7a3c0b80-65cc-11e9-95e9-ff063e1d6547.png)

ViktorHofer commented:
Seems like I'm missing something for the .exe.config copying. Probably a dependent target I need for the manual packaging.
